File: HW2/hw2_problem5.m
%% MAE 6760 Model Based Estimation
% Cornell University
% M Campbell
%
% Homework #2
% Problem #5: Nonlinear Least Squares
%   for multi-sensor range localization problem
%
rng(10);

%% Problem set up with beacons and msmt generation
%Beacon locations
bA = [-10;100];
bB = [490;20];
bC = [500;40];
beacons=[bA bB bC];
%True location of the car
xtrue = [-5;2];

%Perfect measurements
RA = sqrt((bA(1) - xtrue(1))^2 + (bA(2) - xtrue(2))^2);
RB = sqrt((bB(1) - xtrue(1))^2 + (bB(2) - xtrue(2))^2);
RC = sqrt((bC(1) - xtrue(1))^2 + (bC(2) - xtrue(2))^2);

%a total of 10 measurements
n=10;


%% Part (a) - two ranging beacons (A,B,C)
%generate noisy measurements to beacons A,B,C
Rpart_a = diag([10 10 10]);
v_a = sqrtm(Rpart_a)*randn(3,n);
z_a = repmat([RA;RB;RC],[1,n]) + v_a;


%% Part (b) - two ranging beacons (A,B)
%Get noisy measurements to beacons A,B from part (a)
%simply so that the same noise is used
ii_beacons=[1 2];
Rpart_b1 = Rpart_a(ii_beacons,ii_beacons); 
z_b1 = z_a(ii_beacons,:);


%% Part (b) - two ranging beacons (B,C)
%Get noisy measurements to beacons B,C from part (a)
%simply so that the same noise is used
ii_beacons=[2 3];
Rpart_b2 = Rpart_a(ii_beacons,ii_beacons); 
z_b2 = za(ii_beacons,:);
x0=[0;0];


%% Part (c) - perfect linearization

%% Part (d) - correlated noise model
Rpart_d = [10 0 0; 0 10 9; 0 9 10];
v_d = sqrtm(Rpartd)*randn(3,n);
z_d = repmat([RA;RB;RC],[1,n]) + v_d;

File: hw2_problem5.py
"""
MAE 6760 Model Based Estimation
Cornell University
M Campbell

Homework #2
Problem #5: Nonlinear Least Squares
for multi-sensor range localization problem
"""

import numpy as np

np.set_printoptions(precision=5, suppress=True)
np.random.seed(10)

# =================================================
# Problem setup: beacons and measurements
# =================================================
bA = np.array([[-10.0], [100.0]])
bB = np.array([[490.0], [20.0]])
bC = np.array([[500.0], [40.0]])
beacons = np.hstack((bA, bB, bC))

xtrue = np.array([[-5.0], [2.0]])

# Perfect ranges
RA = np.linalg.norm(bA - xtrue)
RB = np.linalg.norm(bB - xtrue)
RC = np.linalg.norm(bC - xtrue)

n = 10  # number of measurements

# =================================================
# Part (a): three beacons (A,B,C)
# =================================================
Rpart_a = np.diag([10.0, 10.0, 10.0])
v_a = np.linalg.cholesky(Rpart_a) @ np.random.randn(3, n)
z_a = np.tile(np.array([[RA], [RB], [RC]]), (1, n)) + v_a


# =================================================
# Part (b): two beacons (A,B) and (B,C)
# =================================================
# (A,B)
ii = [0, 1]
Rpart_b1 = Rpart_a[np.ix_(ii, ii)]
z_b1 = z_a[ii, :]


# (B,C)
ii = [1, 2]
Rpart_b2 = Rpart_a[np.ix_(ii, ii)]
z_b2 = z_a[ii, :]


# =================================================
# Part (c): perfect linearization
# =================================================


# =================================================
# Part (d): correlated noise
# =================================================
Rpart_d = np.array([[10, 0, 0],
                    [0, 10, 9],
                    [0, 9, 10]], dtype=float)

v_d = np.linalg.cholesky(Rpart_d) @ np.random.randn(3, n)
z_d = np.tile(np.array([[RA], [RB], [RC]]), (1, n)) + v_d

